### DISCUSSION TITLE
Very simple init.el does not compile via make init-build

### DISCUSSION THREAD

**Discussion by pedz** [emacscollective/borg#135] [Support] [answered]
I trimmed my init.el down to just
```lisp
(declare-function pedz/org-require "early-init.el" (sym &optional dont-load paths))
(pedz/org-require 'pedz)
```
and I do `make init-build` and this is the output:
```
Initializing drones...done (42 drones in 0.039s)

--- [init.el] ---

Loading /Users/pedz/.config/emacs/init.el (source)...
Debugger entered--Lisp error: (void-function pedz/org-require)
  (pedz/org-require 'pedz)
  eval-buffer(#<buffer  *load*> nil "/Users/pedz/.config/emacs/init.el" nil t)  ; Reading at buffer position 109
  load-with-code-conversion("/Users/pedz/.config/emacs/init.el" "/Users/pedz/.config/emacs/init.el" nil nil)
  load("/Users/pedz/.config/emacs/init.el" nil nil t)
  load-file("init.el")
  borg-batch-rebuild-init()
  command-line-1(("-L" "lib/borg/" "--load" "borg" "--funcall" "borg-initialize" "--funcall" "borg-batch-rebuild-init"))
  command-line()
  normal-top-level()

make: *** [init-build] Error 255
```

But, if I do it by hand with `emacs -Q --batch --eval '(byte-compile-file "init.el")'` it works.

Update: I forgot to mention, I worked around the problem with:
```lisp
(eval-when-compile
  (defun pedz/org-require (sym &optional dont-load paths)))
(pedz/org-require 'pedz)
```

**Comment by tarsius** [comment, marked as answer]
IMO that's not a "very simply init.el", rather you are doing something quite unusual.

If you figure this out, please post the result here. But meanwhile, I have too much on my hands to actively help you investigate this. I don't understand what you are trying to do and why, and it seems like an unnecessary complication to me.

**Comment by pedz** [comment]
To me, I’m just trying to call a function that is declared outside of init.

I’ll poke at it some more and see what I find.

**Comment by pedz** [comment]
Starting with just:
```lisp
(foo-dog "hi")
```
which is just calling a function defined outside of init, it errors out in the same way.

In `borg-batch-rebuild-init` [line 564](https://github.com/emacscollective/borg/blob/master/borg.el#L564) the init file is loaded.  Since the function is not defined, the load fails.  The next line is the call to `byte-recompile-file`.

If I comment out the `load`, then I get the expected
>Warning: the function ‘foo-dog’ is not known to be defined.

I add in the (declare-function ...) and the compile works.

I'm guessing you do the load because you want the things that init loads to be defined?  I'm not sure.

My particular use case, I had a function that kept moving and evolving and I finally put it into early-init.el.  Thus, when Emacs launches normally, it will be available.

In any case, I'm just trying to help.
Take care

**Reply by tarsius** [reply]
> My particular use case, I had a function that kept moving and evolving and I finally put it into early-init.el.

I don't understand why you don't just define the function in the file in which you use it. (But depending on how you use it you might have to wrap it with `eval-and-compile`.)

> Thus, when Emacs launches normally, it will be available.

You could explicitly load the file in which the function is defined. You can make `init.el` explicitly load `early-init.el` at compile time, but if you do that you should probably put some mechanism in place that makes sure it doesn't get loaded twice.

**Reply by tarsius** [reply]
> I'm guessing you do the load because you want the things that init loads to be defined? I'm not sure.

I believe to remember that not doing that results in a lot of warnings from the byte-compiler.
